src/lib/party.ts

```diff
@@ -819,57 +823,48 @@ class Party implements Types.Party {
     return this.body.publicKey;
   }
 
-  signInPlace(privateKey?: PrivateKey, fallbackToZeroNonce = false) {
-    this.setNoncePrecondition(fallbackToZeroNonce);
+  sign(privateKey?: PrivateKey) {
+    let nonce = Party.getNonce(this);
+    this.account.nonce.assertEquals(nonce);
     this.body.incrementNonce = Bool(true);
     Authorization.setLazySignature(this, { privateKey });
   }
 
-  sign(privateKey?: PrivateKey) {
-    let party = Party.clone(this);
-    party.signInPlace(privateKey);
-    return party;
-  }
-
   static signFeePayerInPlace(
     feePayer: FeePayerUnsigned,
-    privateKey?: PrivateKey,
-    fallbackToZeroNonce = false
+    privateKey?: PrivateKey
   ) {
-    feePayer.body.nonce = this.getNonce(feePayer, fallbackToZeroNonce);
+    feePayer.body.nonce = this.getNonce(feePayer);
     feePayer.authorization = Ledger.dummySignature();
     feePayer.lazyAuthorization = { kind: 'lazy-signature', privateKey };
   }
 
-  // TODO this needs to be more intelligent about previous nonces in the transaction, similar to Party.createSigned
```

Comment:
the changes here and in `Party.createSigned` unify the logic of computing the right nonce, and make it correct in more cases (e.g., when child parties exist)